File: Poisson_test/fd2poisson.m
% Numerical approximation to Poisson’s equation over the square [a,b]x[a,b] with
% Dirichlet boundary conditions. Uses a uniform mesh with (n+2)x(n+2) total
% points (i.e, n x n interior grid points).
% Input:
% pfunc : the RHS of poisson equation (i.e. the Laplacian of u).
% bfunc : the boundary function representing the Dirichlet B.C.
% a,b : the interval defining the square
% n : n+2 is the number of points in either direction of the mesh.
% Ouput:
% u : the numerical solution of Poisson equation at the mesh points.
% x,y : the uniform mesh.


function [n2,x,y] = fd2poisson(Lx,Lz,Nx,Nz,Rhs)

% compute the dx and dz
hx = (Lx)/(Nx+1); 
hz = (Lz)/(Nz+1);

[x,y] = meshgrid(0:Nx,0:Nz); % mesh, including boundary points. 

% Compute u on the boundary from the Dirichlet boundary condition
ub = zeros(Nx,Nz);
idx = 2:Nx+1;
idy = 2:Nz+1;

% West and East boundaries need special attention
% ub(:,1) = ; % West Boundary
% ub(:,n) = ; % East Boundary

% Now the North and South boundaries
ub(1,1:Nx) = 1.33;
ub(Nz,1:Nx) = 1.43;
% Convert ub to a vector using column reordering
ub = (1/hx^2)*reshape(ub,Nx*Nz,1);

% Call the RHS of Poisson’s equation at the interior points.
% Convert f to a vector using column reordering
f = reshape(Rhs,Nx*Nz,1);

% Create the D2x and D2y matrices
z = [-2;1;zeros(Nz-2,1)];
D2x = 1/h^2*kron(toeplitz(z,z),eye(Nx));
D2y = 1/h^2*kron(eye(Nz),toeplitz(z,z));

% Solve the system
u = (D2x + D2y)\(f-ub);
% Convert u from a column vector to a matrix to make it easier to work with
% for plotting.
u = reshape(u,Nx,Nz);
% Append on to u the boundary values from the Dirichlet condition.
% u = [[feval(bfunc,x(1,1:n+2),y(1,1:n+2))];...
% [[feval(bfunc,x(2:n+1,1),y(2:n+1,1))] u ...
% [feval(bfunc,x(2:n+1,n+2),y(2:n+1,n+2))]];...
% [feval(bfunc,x(n+2,1:n+2),y(n+2,1:n+2))]];

n2=u;

File: poisson_test_python/fd2poisson.py
import numpy as np
from scipy import sparse
from scipy.sparse.linalg import spsolve
from scipy.linalg import toeplitz

def fd2poisson(Lx, Lz, Nx, Nz, Rhs):
    """
    Numerical approximation to Poisson's equation over the square [0,Lx]x[0,Lz] with
    Dirichlet boundary conditions. Uses a uniform mesh with (Nx+2)x(Nz+2) total
    points (i.e, Nx x Nz interior grid points).

    Parameters:
    -----------
    Lx : float
        Width of the domain
    Lz : float
        Height of the domain
    Nx : int
        Number of interior grid points in x-direction
    Nz : int
        Number of interior grid points in z-direction
    Rhs : ndarray
        Right-hand side of the Poisson equation

    Returns:
    --------
    n2 : ndarray
        Solution of the Poisson equation
    x : ndarray
        x-coordinates of the grid
    y : ndarray
        y-coordinates of the grid
    """
    # Compute the dx and dz
    hx = Lx / (Nx + 1)
    hz = Lz / (Nz + 1)

    # Create mesh, including boundary points
    x, y = np.meshgrid(np.arange(Nx+1), np.arange(Nz+1))

    # Compute u on the boundary from the Dirichlet boundary condition
    ub = np.zeros((Nx, Nz))

    # North and South boundaries
    ub[0, :] = 1.33  # North boundary
    ub[Nz-1, :] = 1.43  # South boundary

    # Convert ub to a vector using column reordering
    ub = (1 / hx**2) * ub.flatten(order='F')

    # Convert f to a vector using column reordering
    f = Rhs.flatten(order='F')

    # Create the D2x and D2y matrices
    z = np.array([-2, 1] + [0] * (Nz - 2))
    D2x = (1 / hx**2) * sparse.kron(sparse.csr_matrix(toeplitz(z, z)), sparse.eye(Nx))
    D2y = (1 / hz**2) * sparse.kron(sparse.eye(Nz), sparse.csr_matrix(toeplitz(z, z)))

    # Solve the system
    u = spsolve(D2x + D2y, f - ub)

    # Convert u from a column vector to a matrix
    u = u.reshape((Nx, Nz), order='F')

    return u, x, y
